### ISSUE TITLE
Observed bookies are getting unhealthy with exception "ERROR o.a.bookkeeper.client.PendingReadOp: Read of ledger entry failed"

### ISSUE THREAD

**Issue by sumit-bm** [pravega/pravega#3523]
While running IO workload from pravega benchmark (Total - 10 readers, 10 writers, 1000 events/sec, 10 streams with 10 segments), and transaction workload with 1 reader and 1 writter observed Bookies become unhealthy after 4 hours or so of IO loading.  

Environment details: PKS / K8 with medium cluster:
```
3 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
5 worker nodes @ xlarge.cpu(8 cpu, 8 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
Pravega details:
```
Pravega version: zk-closed-client-issue-0.5.0-2162.0bbfa42
Zookeeper Operator : pravega/zookeeper-operator:0.2.1
Pravega Operator: pravega/pravega-operator:0.3.1
```

After 4h of transaction IO, there is `RetriesExhaustedException`, and `ConnectionFailedException: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(..) failed: Connection refused: /172.24.0.7:12345`

Now, though all bookies haven't restarted but all are in not healthy condition,
```
Events:
  Type     Reason     Age   From                                           Message
  ----     ------     ----  ----                                           -------
  Warning  Unhealthy  35m   kubelet, d6cfa064-32e8-466e-bbeb-1219b95c8f9f  Readiness probe failed: 03:45:51,191 INFO  Created ledger 22474
03:45:58,151 ERROR Could not get additional bookie to remake ensemble, closing ledger: 22474
03:45:58,151 ERROR Closing ledger 22474 due to NotEnoughBookiesException: Not enough non-faulty bookies available
03:45:58,152 ERROR Write of ledger entry to quorum failed: L22474 E0
03:45:58,153 WARN  Error in bookie sanity test
```
BKs are not replying to any of the segmentstore request:
```
log_pod_pravega-pravega-segmentstore-0_recent.log:2019-03-24 03:23:00,436 23770406 [BookKeeperClientWorker-OrderedExecutor-0-0] ERROR o.a.bookkeeper.client.PendingReadOp - Read of ledger entry failed: L6395 E10201-E12186, Sent to [pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-3.pravega-bookie-headless.default.svc.cluster.local:3181], Heard from [] : bitset = {}. First unread entry is 10201
log_pod_pravega-pravega-segmentstore-0_recent.log:2019-03-24 03:23:00,437 23770407 [core-3] ERROR i.p.s.server.logs.DurableLog - DurableLog[3] Recovery FAILED.
log_pod_pravega-pravega-segmentstore-0_recent.log:2019-03-24 03:23:11,152 23781122 [epollEventLoopGroup-11-2] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection:
```

**Comment by fpj** [comment]
Is this a scenario in which the cluster is unavailable?

**Comment by sumit-bm** [comment]
No, cluster is available but bookies are not responding to Segmentstore request. 
Bookies doesn't showed restarted over here as you see in the o/p but read of ledger entry is failing for all Segmentstores.

**Comment by andreipaduroiu** [comment]
@fpj Since we only read from the bookies during recovery, then whatever container is trying to recover won't be able to serve requests until such reads are successful. So I would say that at least part of the cluster is unavailable.

**Comment by RaulGracia** [comment]
@sumit-bm has this issue been reproduced again? Apart from other issue that have been solved for `0.5`, `NotEnoughBookiesException` may be also due to problems with the communication with Zookeeper. One of the reasons (may not be the only one) for the problems we experienced with Zookeeper is being addressed in https://github.com/pravega/zookeeper-operator/pull/69. Moreover, as this issue is reporting problems just 4 hours after starting the workload, I assume that this is not the case anymore (longevity runs are lasting +5 days). Please, re-open the issue if you encounter this problem again.
